packages/core/src/browser/tree/tree-react-widget.tsx

```diff
@@ -0,0 +1,715 @@
+            this.model.onExpansionChanged(() => this.updateDecorations(this.decoratorService.getDecorations(this.model))),
+            this.decoratorService,
+            this.decoratorService.onDidChangeDecorations(op => this.updateDecorations(op(this.model)))
+        ]);
+        setTimeout(() => this.updateDecorations(this.decoratorService.getDecorations(this.model)), 0);
+    }
+
+    protected async updateDecorations(decorations: MaybePromise<Map<string, TreeDecoration.Data[]>>): Promise<void> {
+        this.decorations = await decorations;
+        this.update();
+    }
+
+    protected onActivateRequest(msg: Message): void {
+        super.onActivateRequest(msg);
+        this.node.focus();
+        if (this.model.selectedNodes.length === 0) {
+            const root = this.model.root;
+            if (SelectableTreeNode.is(root)) {
+                this.model.selectNode(root);
+            } else if (CompositeTreeNode.is(root) && root.children.length >= 1) {
+                const firstChild = root.children[0];
+                if (SelectableTreeNode.is(firstChild)) {
+                    this.model.selectNode(firstChild);
+                }
+            }
+        }
+    }
+
+    protected onUpdateRequest(msg: Message): void {
+        super.onUpdateRequest(msg);
+
+        const focus = this.node.getElementsByClassName(FOCUS_CLASS)[0];
+        if (focus) {
+            ElementExt.scrollIntoViewIfNeeded(this.node, focus);
+        } else {
+            const selected = this.node.getElementsByClassName(SELECTED_CLASS)[0];
+            if (selected) {
+                ElementExt.scrollIntoViewIfNeeded(this.node, selected);
+            }
+        }
+    }
+
+    protected render(): React.ReactNode {
+        return React.createElement("div", this.createContainerAttributes(), this.renderTree(this.model));
+    }
+
+    protected createContainerAttributes(): React.HTMLAttributes<HTMLElement> {
+        return {
+            className: TREE_CONTAINER_CLASS,
+            onContextMenu: e => this.handleContextMenuEvent(this.model.root, e)
+        };
+    }
+
+    protected onAfterAttach(msg: Message): void {
+        const up = [
+            Key.ARROW_UP,
+            KeyCode.createKeyCode({ first: Key.ARROW_UP, modifiers: [KeyModifier.Shift] })
+        ];
+        const down = [
+            Key.ARROW_DOWN,
+            KeyCode.createKeyCode({ first: Key.ARROW_DOWN, modifiers: [KeyModifier.Shift] })
+        ];
+        super.onAfterAttach(msg);
+        this.addKeyListener(this.node, Key.ARROW_LEFT, event => this.handleLeft(event));
+        this.addKeyListener(this.node, Key.ARROW_RIGHT, event => this.handleRight(event));
+        this.addKeyListener(this.node, up, event => this.handleUp(event));
+        this.addKeyListener(this.node, down, event => this.handleDown(event));
+        this.addKeyListener(this.node, Key.ENTER, event => this.handleEnter(event));
```

Comment:
We had trouble with mixing up native and react events. So the event is now registered [here](https://github.com/theia-ide/theia/pull/2224/files#diff-af560e340c6ed227bc46a9abc08c8912R178).

Reply:
Thanks for the link and the explanation. Can you please give me some insight into this _trouble_?

Reply:
React delegates every event which is registered on a react element to the document. Here are the handlers stored and called. So if we register a handler on some tree node via react and register a handler for the same event on the root node not via react (like in that old case) the handler of the root node is processed. And because of the `stopPropagation` the event does not reach the document root and the handlers of the tree nodes aren't called. 
That is how I understood this situation. 
More about that: https://github.com/facebook/react/issues/7094#issuecomment-228931737
...and: https://davidwalsh.name/event-delegate

Reply:
I need to try this out. Thanks!